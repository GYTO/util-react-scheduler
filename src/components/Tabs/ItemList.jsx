// @flow
import * as React from 'react';
import { PRODUCTS } from "../../api/producs";
import type { Products } from "../types/Products";

type Props = {
    onClick: (index: number) => void,
}

type State = {
    products: Products[]
}

const boxStyle = {
    border: '1px solid gray',
    width: '200px',
    display: 'flex',
};

const boxes = {
    display: 'flex',
    flexDirection: 'column',
};

class ItemList extends React.Component<Props, State> {
    constructor(props: Props) {
        super(props);

        this.state = {
            products: PRODUCTS,
        };
    }

    render() {
        const { products } = this.state;
        const { onClick } = this.props;
        return (
            <div style={boxes}>
                {products.map((item, index) => {
                    return <div key={index} style={boxStyle} onClick={() => onClick(item.id)}>
                        <img src={item.img} alt={item.name} />
                        <h4>{item.name}</h4>
                    </div>;
                })}
            </div>
        );
    }
}

export default ItemList;
