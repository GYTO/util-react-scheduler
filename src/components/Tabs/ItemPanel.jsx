// @flow
import * as React from 'react';
import { PRODUCT_ITEM} from '../../api/producs';
import type { ProductItem } from '../types/Products';

type Props = {
    selected: ?number
}

type State = {
    productItem: ProductItem[],
}

class ItemPanel extends React.Component<Props, State> {
    constructor(props: Props) {
        super(props);

        this.state = {
            productItem: PRODUCT_ITEM,
        };
    }

    render() {
        const { productItem } = this.state;
        const selectedItem = productItem.filter(item => {
            return item.id === this.props.selected;
        });

        return (
            <div>
                {this.props.selected ? selectedItem.map((item, index) => {
                    return <div key={index}>
                        <p>{item.id}</p>
                        <p>{item.name}</p>
                        <p>{item.text}</p>
                    </div>;
                }) : <p>Please select product tab</p>}
            </div>
        );
    }
}

export default ItemPanel;
