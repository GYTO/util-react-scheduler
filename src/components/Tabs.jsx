// @flow
import * as React from 'react';
import ItemList from './Tabs/ItemList';
import ItemPanel from './Tabs/ItemPanel';

type Props = {}

type State = {
    selectedProduct: ?number
}

const panel = {
    display: 'flex',
};

class Tabs extends React.Component<Props, State> {
    constructor(props: Props) {
        super(props);

        this.state = {
            selectedProduct: null,
        };
    }

    openTab = (index: number) => {
        this.setState({ selectedProduct: index});
    };

    render() {
        const { selectedProduct } = this.state;
        return (
            <div style={panel}>
                <ItemList itemClick={this.openTab}/>
                <ItemPanel selected={selectedProduct}/>
            </div>
        );
    }
}

export default Tabs;
