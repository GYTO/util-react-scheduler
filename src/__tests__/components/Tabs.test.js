import * as React from 'react';
import { mount, shallow } from 'enzyme';
import Tabs from '../../components/Tabs';

class MockItemList extends React.Component<{}> {
    render() {
        return <div/>;
    }
}
jest.mock('../../components/Tabs/ItemList', () => {
    return jest.fn().mockImplementation((passedProps) => {
        return <MockItemList {...passedProps}/>;
    });
});

class MockItemPanel extends React.Component<{}> {
    render() {
        return <div/>;
    }
}
jest.mock('../../components/Tabs/ItemPanel', () => {
    return jest.fn().mockImplementation((passedProps) => {
        return <MockItemPanel {...passedProps}/>;
    });
});

describe('Tabs', () => {
    let props;
    let mountedRender;
    let shallowRender;

    const getMounted = () => {
        if (!mountedRender) {
            mountedRender = mount(
                <Tabs {...props} />
            );
        }
        return mountedRender;
    };

    const getShallow = () => {
        if (!shallowRender) {
            shallowRender = shallow(
                <Tabs {...props} />
            );
        }
        return shallowRender;
    };

    beforeEach(() => {
        jest.clearAllMocks();

        mountedRender = undefined;
    });

    it('matches snapshot', () => {
        expect(getShallow()).toMatchSnapshot();
    });

    it('expecting MockItemList prop onClick to be function', () => {
        expect(getMounted().find('MockItemList')).toHaveLength(1);
        expect(typeof getMounted().find('MockItemList').prop('onClick')).toBe('function');
    });

    it('should change a prop of MockItemPanel when MockItemList is clicked', () => {
        getMounted().find('MockItemList').prop('onClick')(1);
        getMounted().update();

        expect(getMounted().find('MockItemPanel').prop('selected')).toEqual(1);
    });
});
