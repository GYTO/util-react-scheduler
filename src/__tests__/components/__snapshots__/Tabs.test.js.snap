// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs matches snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Tabs />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <mockConstructor
          onClick={[Function]}
        />,
        <mockConstructor
          selected={null}
        />,
      ],
      "style": Object {
        "display": "flex",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "onClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [MockFunction],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "selected": null,
        },
        "ref": null,
        "rendered": null,
        "type": [MockFunction],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <mockConstructor
            onClick={[Function]}
          />,
          <mockConstructor
            selected={null}
          />,
        ],
        "style": Object {
          "display": "flex",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [MockFunction],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "selected": null,
          },
          "ref": null,
          "rendered": null,
          "type": [MockFunction],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
